### ISSUE TITLE
[IMPROVEMENT] Investigate how Longhorn is working with CAPI rolling upgrade

### ISSUE THREAD

**Issue by PhanLe1010** [longhorn/longhorn#8362]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

We would like to investigate how Longhorn behaves in CAPI rolling upgrade case. How to make Longhorn reuse the replica in during this process instead of creating new replicas?

This reproducing steps is copied from the ticket https://gitlab.com/sylva-projects/sylva-core/-/issues/1141#reproduction-steps 

### Reproduction steps:

1. Deploy Sylva with longhorn with 3 BM nodes.
2. Set `replica-replenishment-wait-interval=80 minutes` for longhorn and `numberOfReplicas: "2"`
3. After deployment is completed, note down the details of longhorn replicas in GUI like name , state and status.
4. Before starting rolling upgrade, set below values in environmental-values
    ```
    kubelet_extra_args:
        max-pods: "228" 
    capm3:
        nodeReuse: true
        automatedCleaningMode: disabled
    ```
5. Start rolling upgrade.
6. Check the status of longhorn replicas after rolling upgrade is completed.

## Expected behavior:
* Longhorn should not create new replica and should reuse existing (i.e. old) replica
* Longhorn should rebuild the data using existing (i.e. old) replica

**Comment by derekbit** [comment]
cc @c3y1huang @ejweber

**Comment by mantissahz** [comment]
Related: https://github.com/metal3-io/cluster-api-provider-metal3/issues/1584

**Comment by Martin-Weiss** [comment]
Basically we need these things:

1. upgrade kubernetes and keeping the longhorn data on /var/lib/longhorn or an additional disk (this is covered in https://longhorn.io/docs/1.6.0/maintenance/maintenance/#in-place-upgrade )
2. redeploy the OS while keeping the node-id/node-name and keeping the longhorn data on /var/lib/longhorn or an additional disk
3. redeploy the OS while getting new node-id/node-name and keeping the longhorn data on /var/lib/longhorn or an additional disk

I am not sure if https://github.com/longhorn/longhorn/issues/8362 will be for 2. or for 3.?
Basically I assume that the challenge might be to "identify and re-use existing replicas in case the node-id/node-name changes"?

**Comment by PhanLe1010** [comment]
Additional info from @Martin-Weiss 

> As far as I understood - the major problem is that from a cloud-native approach an upgrade is always "throw away a node and create a new node" which per default will create new node-id/node-name - so if "3" is missing - I guess longhorn might need some design adjustments for "cloud-native" ? :wink: - basically it would be a good feature if we could move any longhorn data-disk from one node to another and it re-detects and re-uses the data automatically..
>
> also 3 can be seen as a "disaster recovery" scenario as well, where the node is destroyed for whatever reason but you can still can use the longhorn device into a fresh one

**Comment by guettli** [comment]
@PhanLe1010 How did you solve that? We use cluster-api and face the same problems. I created an idea how this could be solved in an automated way: #8962 What do you think about that?

**Comment by rpardini** [comment]
This really depends on details of the specific Cluster API implementations used, and the way node replacements are done ref naming, and the way longhorn data disk is or isn't preserved across reprovisionings.

There's two scenarios where this already works. Both require `nodeDrainPolicy: block-for-eviction-if-contains-last-replica` or stronger. I assume the default Longhorn disk, `/var/lib/longhorn`.

1) If Nodes are **replaced with nodes with new (unique) names**, and `/var/lib/longhorn` is **not** preserved; this works nicely: CAPI drains will block until Longhorn evicts the replicas on the node. When a new Node joins (with a new Node name, and empty `/var/lib/longhorn`, Longhorn will see it as a completely new Node/Disk. The only detail here is that over time, Longhorn will accumulate a long list of "Down" Nodes (with the old names) which are still `schedulable: true` and thus won't be cleaned up. 

2) If Nodes are replaced with new Nodes with the **same name as before**, and `/var/lib/longhorn` is a mount for a filesystem that is **fully preserved (including it's UUID)**: works. Longhorn is blissfully unaware of any Node "replacement", and to its eyes, all that happened was a node "reboot"/"upgrade".

AFAIK, any deviation from those two scenarios won't work. Example: If Nodes are replaced with same name, but `/var/lib/longhorn` either isn't preserved, or it's UUID changes, Longhorn won't accept it back without manual intervention (one needs to delete the old Disk manually).

**Comment by COLDTURNIP** [comment]
Exactly node replacement is complicated. The volume detaching is triggered by node draining of CAPI Machine deletion. Maybe we could have an operator in the CAPI manager cluster, registers the Machine pre-termination hook (e.g. `pre-terminate.delete.hook.machine.cluster.x-k8s.io/longhorn-node-eviction`), and remotely trigger LHN eviction in workload cluster when Machine deletion is triggered.
